File: android/src/main/java/com/gameanalytics/sdk/flutter/GameAnalyticsPlugin.java
package com.gameanalytics.sdk.flutter;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.gameanalytics.sdk.GameAnalytics;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** GameAnalyticsPlugin */
public class GameAnalyticsPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware
{
    private MethodChannel channel;
    private Activity activity;
    private static final String VERSION = "1.2.4";

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding)
    {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "gameanalytics");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result)
    {
        if (call.method.equals("configureAvailableCustomDimensions01"))
        {
            String customDimensions = call.argument("customDimensions");

            try
            {
                JSONArray jsonArray = new JSONArray(customDimensions);

                ArrayList<String> array = new ArrayList<>();

                for(int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }

                GameAnalytics.configureAvailableCustomDimensions01(array.toArray(new String[0]));
            }
            catch (JSONException e)
            {
                e.printStackTrace();
            }
        }
        else if (call.method.equals("configureAvailableCustomDimensions02"))
        {
            String customDimensions = call.argument("customDimensions");

            try
            {
                JSONArray jsonArray = new JSONArray(customDimensions);

                ArrayList<String> array = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }

                GameAnalytics.configureAvailableCustomDimensions02(array.toArray(new String[0]));
            }
            catch (JSONException e)
            {
                e.printStackTrace();
            }
        }
        else if (call.method.equals("configureAvailableCustomDimensions03"))
        {
            String customDimensions = call.argument("customDimensions");

            try
            {
                JSONArray jsonArray = new JSONArray(customDimensions);

                ArrayList<String> array = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }

                GameAnalytics.configureAvailableCustomDimensions03(array.toArray(new String[0]));
            }
            catch (JSONException e)
            {
                e.printStackTrace();
            }
        }
        else if (call.method.equals("configureAvailableResourceCurrencies"))
        {
            String resourceCurrencies = call.argument("resourceCurrencies");

            try
            {
                JSONArray jsonArray = new JSONArray(resourceCurrencies);

                ArrayList<String> array = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }

                GameAnalytics.configureAvailableResourceCurrencies(array.toArray(new String[0]));
            }
            catch (JSONException e)
            {
                e.printStackTrace();
            }
        }
        else if (call.method.equals("configureAvailableResourceItemTypes"))
        {
            String resourceItemTypes = call.argument("resourceItemTypes");

            try
            {
                JSONArray jsonArray = new JSONArray(resourceItemTypes);

                ArrayList<String> array = new ArrayList<>();

                for (int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }

                GameAnalytics.configureAvailableResourceItemTypes(array.toArray(new String[0]));
            }
            catch (JSONException e)
            {
                e.printStackTrace();
            }
        }
        else if (call.method.equals("configureBuild"))
        {
            String build = call.argument("build");
            GameAnalytics.configureBuild(build);
        }
        else if (call.method.equals("configureAutoDetectAppVersion"))
        {
            Boolean flag = call.argument("flag");
            if(flag != null)
            {
                GameAnalytics.configureAutoDetectAppVersion(flag);
            }
            else
            {
                result.error("ERROR", "configureAutoDetectAppVersion: Flag argument could not be parsed to a boolean", null);
            }
        }
        else if (call.method.equals("configureUserId"))
        {
            String uId = call.argument("uId");
            GameAnalytics.configureUserId(uId);
        }
        else if (call.method.equals("initialize"))
        {
            String gameKey = call.argument("gameKey");
            String secretKey = call.argument("secretKey");
            String sdkVersion = "flutter " + VERSION;
            GameAnalytics.configureSdkGameEngineVersion(sdkVersion);
            GameAnalytics.initialize(this.activity, gameKey, secretKey);
        }
        else if (call.method.equals("addBusinessEvent"))
        {
            String currency = call.argument("currency");
            Integer amount = call.argument("amount");
            String itemType = call.argument("itemType");
            String itemId = call.argument("itemId");
            String cartType = call.argument("cartType");
            String fields = "{}";
            if (call.hasArgument("customFields"))
            {
                fields = call.argument("customFields");
            }
            boolean mergeFields = false;
            if (call.hasArgument("mergeFields")) {
                Boolean b = call.argument("mergeFields");
                mergeFields = b.booleanValue();
            }

            if(amount != null)
            {
                if(call.hasArgument("receipt") && call.hasArgument("signature"))
                {
                    String receipt = call.argument("receipt");
                    String signature = call.argument("signature");
                    GameAnalytics.addBusinessEvent(currency, amount, itemType, itemId, cartType, receipt, "google_play", signature,
                            fields, mergeFields);
                }
                else
                {
                    GameAnalytics.addBusinessEvent(currency, amount, itemType, itemId, cartType, fields, mergeFields);
                }
            }
            else
            {
                result.error("ERROR", "addBusinessEvent: Amount argument could not be parsed to an integer", null);
            }
        }
        else if (call.method.equals("addResourceEvent"))
        {
            Integer flowType = call.argument("flowType");
            String currency = call.argument("currency");
            Number amount = call.argument("amount");
            String itemType = call.argument("itemType");
            String itemId = call.argument("itemId");
            String fields = "{}";
            if (call.hasArgument("customFields")) {
                fields = call.argument("customFields");
            }
            boolean mergeFields = false;
            if (call.hasArgument("mergeFields")) {
                Boolean b = call.argument("mergeFields");
                mergeFields = b.booleanValue();
            }

            if(flowType != null && amount != null)
            {
                GameAnalytics.addResourceEvent(flowType, currency, amount.floatValue(), itemType, itemId, fields,
                        mergeFields);
            }
            else
            {
                result.error("ERROR", "addResourceEvent: flowType or amount argument could not be parsed to the correct type", null);
            }
        }
        else if (call.method.equals("addProgressionEvent"))
        {
            Integer progressionStatus = call.argument("progressionStatus");
            String progression01 = call.argument("progression01");
            String progression02 = "";
            if(call.hasArgument("progression02"))
            {
                progression02 = call.argument("progression02");
            }
            String progression03 = "";
            if (call.hasArgument("progression03"))
            {
                progression03 = call.argument("progression03");
            }
            Number score = null;
            boolean sendScore = false;
            if (call.hasArgument("score"))
            {
                score = call.argument("score");
                sendScore = true;
            }
            String fields = "{}";
            if (call.hasArgument("customFields")) {
                fields = call.argument("customFields");
            }
            boolean mergeFields = false;
            if (call.hasArgument("mergeFields")) {
                Boolean b = call.argument("mergeFields");
                mergeFields = b.booleanValue();
            }

            if(progressionStatus != null)
            {
                if (sendScore && score != null)
                {
                    GameAnalytics.addProgressionEvent(progressionStatus, progression01, progression02, progression03, score.doubleValue(), fields,
                            mergeFields);
                }
                else
                {
                    GameAnalytics.addProgressionEvent(progressionStatus, progression01, progression02, progression03,
                            fields, mergeFields);
                }
            }
            else
            {
                result.error("ERROR", "addProgressionEvent: progressionStatus argument could not be parsed to an integer", null);
            }
        }
        else if (call.method.equals("addDesignEvent"))
        {
            String eventId = call.argument("eventId");
            Number value = null;
            boolean sendValue = false;
            if (call.hasArgument("value"))
            {
                value = call.argument("value");
                sendValue = true;
            }
            String fields = "{}";
            if (call.hasArgument("customFields")) {
                fields = call.argument("customFields");
            }
            boolean mergeFields = false;
            if (call.hasArgument("mergeFields")) {
                Boolean b = call.argument("mergeFields");
                mergeFields = b.booleanValue();
            }

            if (sendValue && value != null)
            {
                GameAnalytics.addDesignEvent(eventId, value.doubleValue(), fields, mergeFields);
            }
            else
            {
                GameAnalytics.addDesignEvent(eventId, fields, mergeFields);
            }
        }
        else if (call.method.equals("addErrorEvent"))
        {
            Integer severity = call.argument("severity");
            String message = call.argument("message");
            String fields = "{}";
            if (call.hasArgument("customFields")) {
                fields = call.argument("customFields");
            }
            boolean mergeFields = false;
            if (call.hasArgument("mergeFields")) {
                Boolean b = call.argument("mergeFields");
                mergeFields = b.booleanValue();
            }

            if(severity != null)
            {
                GameAnalytics.addErrorEvent(severity, message, fields, mergeFields);
            }
            else
            {
                result.error("ERROR", "addErrorEvent: severity argument could not be parsed to an integer", null);
            }
        }
        else if (call.method.equals("addAdEvent"))
        {
            Integer adAction = call.argument("adAction");
            Integer adType = call.argument("adType");
            String adSdkName = call.argument("adSdkName");
            String adPlacement = call.argument("adPlacement");
            Long duration = null;
            boolean sendDuration = false;
            if (call.hasArgument("duration"))
            {
                duration = call.argument("duration");
                sendDuration = true;
            }
            Integer noAdReason = 0;
            boolean sendNoAdReason = false;
            if (call.hasArgument("noAdReason"))
            {
                noAdReason = call.argument("noAdReason");
                sendNoAdReason = true;
            }
            String fields = "{}";
            if (call.hasArgument("customFields")) {
                fields = call.argument("customFields");
            }
            boolean mergeFields = false;
            if (call.hasArgument("mergeFields")) {
                Boolean b = call.argument("mergeFields");
                mergeFields = b.booleanValue();
            }

            if(adAction != null && adType != null)
            {
                if (sendDuration && duration != null)
                {
                    GameAnalytics.addAdEvent(adAction, adType, adSdkName, adPlacement, duration, fields, mergeFields);
                }
                else if (sendNoAdReason && noAdReason != null)
                {
                    GameAnalytics.addAdEvent(adAction, adType, adSdkName, adPlacement, noAdReason, fields, mergeFields);
                }
                else
                {
                    GameAnalytics.addAdEvent(adAction, adType, adSdkName, adPlacement, fields, mergeFields);
                }
            }
            else
            {
                result.error("ERROR", "addAdEvent: adAction or adType argument could not be parsed to the correct type", null);
            }
        }
        else if (call.method.equals("setEnabledInfoLog"))
        {
            Boolean flag = call.argument("flag");

            if(flag != null)
            {
                GameAnalytics.setEnabledInfoLog(flag);
            }
            else
            {
                result.error("ERROR", "setEnabledInfoLog: Flag argument could not be parsed to a boolean", null);
            }
        }
        else if (call.method.equals("setEnabledVerboseLog"))
        {
            Boolean flag = call.argument("flag");

            if(flag != null)
            {
                GameAnalytics.setEnabledVerboseLog(flag);
            }
            else
            {
                result.error("ERROR", "setEnabledVerboseLog: Flag argument could not be parsed to a boolean", null);
            }
        }
        else if (call.method.equals("setEnabledManualSessionHandling"))
        {
            Boolean flag = call.argument("flag");

            if(flag != null)
            {
                GameAnalytics.setEnabledManualSessionHandling(flag);
            }
            else
            {
                result.error("ERROR", "setEnabledManualSessionHandling: Flag argument could not be parsed to a boolean", null);
            }
        }
        else if (call.method.equals("setEnabledEventSubmission"))
        {
            Boolean flag = call.argument("flag");

            if(flag != null)
            {
                GameAnalytics.setEnabledEventSubmission(flag);
            }
            else
            {
                result.error("ERROR", "setEnabledEventSubmission: Flag argument could not be parsed to a boolean", null);
            }
        }
        else if (call.method.equals("setCustomDimension01"))
        {
            String dimension = call.argument("dimension");
            GameAnalytics.setCustomDimension01(dimension);
        }
        else if (call.method.equals("setCustomDimension02"))
        {
            String dimension = call.argument("dimension");
            GameAnalytics.setCustomDimension02(dimension);
        }
        else if (call.method.equals("setCustomDimension03"))
        {
            String dimension = call.argument("dimension");
            GameAnalytics.setCustomDimension03(dimension);
        }
        else if (call.method.equals("setGlobalCustomEventFields"))
        {
            String customFields = call.argument("customFields");
            GameAnalytics.setGlobalCustomEventFields(customFields);
        }
        else if (call.method.equals("startSession"))
        {
            GameAnalytics.startSession();
        }
        else if (call.method.equals("endSession"))
        {
            GameAnalytics.endSession();
        }
        else if (call.method.equals("getRemoteConfigsValueAsString"))
        {
            String key = call.argument("key");
            String defaultValue = null;

            if (call.hasArgument("defaultValue"))
            {
                defaultValue = call.argument("defaultValue");
            }

            String returnValue;
            if (defaultValue != null)
            {
                returnValue = GameAnalytics.getRemoteConfigsValueAsString(key, defaultValue);
            }
            else
            {
                returnValue = GameAnalytics.getRemoteConfigsValueAsString(key);
            }
            result.success(returnValue);
        }
        else if (call.method.equals("isRemoteConfigsReady"))
        {
            result.success(GameAnalytics.isRemoteConfigsReady());
        }
        else if (call.method.equals("getRemoteConfigsContentAsString"))
        {
            result.success(GameAnalytics.getRemoteConfigsContentAsString());
        }
        else if (call.method.equals("getABTestingId"))
        {
            result.success(GameAnalytics.getABTestingId());
        }
        else if (call.method.equals("getABTestingVariantId"))
        {
            result.success(GameAnalytics.getABTestingVariantId());
        }
        else
        {
            result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding)
    {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding)
    {
        this.activity = binding.getActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges()
    {

    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding)
    {

    }

    @Override
    public void onDetachedFromActivity()
    {
        this.activity = null;
    }
}

File: ios/Classes/GameAnalyticsPlugin.m
#import "GameAnalyticsPlugin.h"
#import <GameAnalytics/GameAnalytics.h>

#define VERSION @"1.2.7"

@implementation GameAnalyticsPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"gameanalytics" binaryMessenger:[registrar messenger]];
    GameAnalyticsPlugin* instance = [[GameAnalyticsPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"getPlatformVersion" isEqualToString:call.method])
    {
        result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    }
    else if ([@"configureAvailableCustomDimensions01" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *customDimensions = args[@"customDimensions"];
        NSArray *list_array = nil;
        if (customDimensions) {
            list_array = [NSJSONSerialization JSONObjectWithData:[customDimensions dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        [GameAnalytics configureAvailableCustomDimensions01:list_array];
    }
    else if ([@"configureAvailableCustomDimensions02" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *customDimensions = args[@"customDimensions"];
        NSArray *list_array = nil;
        if (customDimensions) {
            list_array = [NSJSONSerialization JSONObjectWithData:[customDimensions dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        [GameAnalytics configureAvailableCustomDimensions02:list_array];
    }
    else if ([@"configureAvailableCustomDimensions03" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *customDimensions = args[@"customDimensions"];
        NSArray *list_array = nil;
        if (customDimensions) {
            list_array = [NSJSONSerialization JSONObjectWithData:[customDimensions dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        [GameAnalytics configureAvailableCustomDimensions03:list_array];
    }
    else if ([@"configureAvailableResourceCurrencies" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *resourceCurrencies = args[@"resourceCurrencies"];
        NSArray *list_array = nil;
        if (resourceCurrencies) {
            list_array = [NSJSONSerialization JSONObjectWithData:[resourceCurrencies dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        [GameAnalytics configureAvailableResourceCurrencies:list_array];
    }
    else if ([@"configureAvailableResourceItemTypes" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *resourceItemTypes = args[@"resourceItemTypes"];
        NSArray *list_array = nil;
        if (resourceItemTypes) {
            list_array = [NSJSONSerialization JSONObjectWithData:[resourceItemTypes dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        [GameAnalytics configureAvailableResourceItemTypes:list_array];
    }
    else if ([@"configureBuild" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *build = args[@"build"];
        if(build == nil)
        {
            build = @"";
        }
        [GameAnalytics configureBuild:build];
    }
    else if ([@"configureAutoDetectAppVersion" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        BOOL flag = [args[@"flag"] boolValue];
        [GameAnalytics configureAutoDetectAppVersion:flag];
    }
    else if ([@"configureUserId" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *uId = args[@"uId"];
        if(uId == nil)
        {
            uId = @"";
        }
        [GameAnalytics configureUserId:uId];
    }
    else if ([@"initialize" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *gameKey = args[@"gameKey"];
        NSString *secretKey = args[@"secretKey"];
        if(gameKey == nil)
        {
            gameKey = @"";
        }
        if(secretKey == nil)
        {
            secretKey = @"";
        }
        NSString* sdkVersion = [NSString stringWithFormat:@"flutter %@", VERSION];

        [GameAnalytics configureSdkVersion:sdkVersion];
        [GameAnalytics initializeWithGameKey:gameKey gameSecret:secretKey];
    }
    else if ([@"addBusinessEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *currency = args[@"currency"];
        NSInteger amount = [args[@"amount"] intValue];
        NSString *itemType = args[@"itemType"];
        NSString *itemId = args[@"itemId"];
        NSString *cartType = args[@"cartType"];
        NSString* receipt = args[@"receipt"];
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        BOOL mergeFields = NO;
        if (args[@"mergeFields"])
        {
            mergeFields = [args[@"mergeFields"] boolValue];
        }

        if(receipt == nil)
        {
            [GameAnalytics addBusinessEventWithCurrency:currency amount:amount itemType:itemType itemId:itemId cartType:cartType autoFetchReceipt:YES customFields:fields_dict mergeFields:mergeFields];
        }
        else
        {
            [GameAnalytics addBusinessEventWithCurrency:currency amount:amount itemType:itemType itemId:itemId cartType:cartType receipt:receipt customFields:fields_dict mergeFields:mergeFields];
        }
    }
    else if ([@"addResourceEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSInteger flowType = [args[@"flowType"] intValue];
        NSString *currency = args[@"currency"];
        NSNumber *amount = args[@"amount"];
        NSString *itemType = args[@"itemType"];
        NSString *itemId = args[@"itemId"];
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        BOOL mergeFields = NO;
        if (args[@"mergeFields"])
        {
            mergeFields = [args[@"mergeFields"] boolValue];
        }

        [GameAnalytics addResourceEventWithFlowType:(GAResourceFlowType)flowType currency:currency amount:amount itemType:itemType itemId:itemId customFields:fields_dict mergeFields:mergeFields];
    }
    else if ([@"addProgressionEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;

        NSInteger progressionStatus = [args[@"progressionStatus"] intValue];
        NSString* progression01 = args[@"progression01"];
        NSString* progression02 = args[@"progression02"];
        NSString* progression03 = args[@"progression03"];
        NSInteger score = 0;
        BOOL sendScore = NO;
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        BOOL mergeFields = NO;
        if (args[@"mergeFields"])
        {
            mergeFields = [args[@"mergeFields"] boolValue];
        }

        if(args[@"score"])
        {
            score = [args[@"score"] intValue];
            sendScore = YES;
        }

        if(sendScore)
        {
            [GameAnalytics addProgressionEventWithProgressionStatus:(GAProgressionStatus)progressionStatus progression01:progression01 progression02:progression02 progression03:progression03 score:score customFields:fields_dict mergeFields:mergeFields];
        }
        else
        {
            [GameAnalytics addProgressionEventWithProgressionStatus:(GAProgressionStatus)progressionStatus progression01:progression01 progression02:progression02 progression03:progression03 customFields:fields_dict mergeFields:mergeFields];
        }
    }
    else if ([@"addDesignEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;

        NSString* eventId = args[@"eventId"];
        NSNumber* value = nil;
        BOOL sendValue = NO;
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        BOOL mergeFields = NO;
        if (args[@"mergeFields"])
        {
            mergeFields = [args[@"mergeFields"] boolValue];
        }

        if(args[@"value"])
        {
            value = args[@"value"];
            sendValue = YES;
        }

        if(sendValue)
        {
            [GameAnalytics addDesignEventWithEventId:eventId value:value customFields:fields_dict mergeFields:mergeFields];
        }
        else
        {
            [GameAnalytics addDesignEventWithEventId:eventId customFields:fields_dict mergeFields:mergeFields];
        }
    }
    else if ([@"addErrorEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;

        NSInteger severity = [args[@"severity"] intValue];
        NSString* message = args[@"message"];
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        BOOL mergeFields = NO;
        if (args[@"mergeFields"])
        {
            mergeFields = [args[@"mergeFields"] boolValue];
        }

        [GameAnalytics addErrorEventWithSeverity:(GAErrorSeverity)severity message:message customFields:fields_dict mergeFields:mergeFields];
    }
    else if ([@"addAdEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;

        NSInteger adAction = [args[@"adAction"] intValue];
        NSInteger adType = [args[@"adType"] intValue];
        NSString* adSdkName = args[@"adSdkName"];
        NSString* adPlacement = args[@"adPlacement"];
        NSInteger duration = 0;
        BOOL sendDuration = NO;
        NSInteger noAdReason = 0;
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        BOOL mergeFields = NO;
        if (args[@"mergeFields"])
        {
            mergeFields = [args[@"mergeFields"] boolValue];
        }

        if(args != nil)
        {
            adAction = args[@"adAction"] && [args[@"adAction"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"adAction"] intValue] : 0;
            adType = args[@"adType"] && [args[@"adType"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"adType"] intValue] : 0;
            adSdkName = args[@"adSdkName"] && [args[@"adSdkName"] isKindOfClass:[NSString class]] ? args[@"adSdkName"] : @"";
            adPlacement = args[@"adPlacement"] && [args[@"adPlacement"] isKindOfClass:[NSString class]] ? args[@"adPlacement"] : @"";
            duration = args[@"duration"] && [args[@"duration"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"duration"] intValue] : 0;
            sendDuration = args[@"duration"] && [args[@"duration"] isKindOfClass:[NSNumber class]] ? YES : NO;
            noAdReason = args[@"noAdReason"] && [args[@"noAdReason"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"noAdReason"] intValue] : 0;
        }

        if(sendDuration)
        {
            [GameAnalytics addAdEventWithAction:(GAAdAction)adAction adType:(GAAdType)adType adSdkName:adSdkName adPlacement:adPlacement duration:duration customFields:fields_dict mergeFields:mergeFields];
        }
        else
        {
            [GameAnalytics addAdEventWithAction:(GAAdAction)adAction adType:(GAAdType)adType adSdkName:adSdkName adPlacement:adPlacement noAdReason:(GAAdError)noAdReason customFields:fields_dict mergeFields:mergeFields];
        }
    }

    else if ([@"addImpressionMaxEvent" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;

        NSString* adNetworkVersion = args[@"version"];

        NSDictionary *fields_dict = nil;
        if (args[@"fields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"fields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }

        [GameAnalytics addImpressionMaxEventWithAdNetworkVersion:adNetworkVersion impressionData:fields_dict];
    }

    else if ([@"setEnabledInfoLog" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        BOOL flag = [args[@"flag"] boolValue];
        [GameAnalytics setEnabledInfoLog:flag];
    }
    else if ([@"setEnabledVerboseLog" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        BOOL flag = [args[@"flag"] boolValue];
        [GameAnalytics setEnabledVerboseLog:flag];
    }
    else if ([@"setEnabledManualSessionHandling" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        BOOL flag = [args[@"flag"] boolValue];
        [GameAnalytics setEnabledManualSessionHandling:flag];
    }
    else if ([@"setEnabledEventSubmission" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        BOOL flag = [args[@"flag"] boolValue];
        [GameAnalytics setEnabledEventSubmission:flag];
    }
    else if ([@"setCustomDimension01" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *dimension = args[@"dimension"];
        if(dimension == nil)
        {
            dimension = @"";
        }
        [GameAnalytics setCustomDimension01:dimension];
    }
    else if ([@"setCustomDimension02" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *dimension = args[@"dimension"];
        if(dimension == nil)
        {
            dimension = @"";
        }
        [GameAnalytics setCustomDimension02:dimension];
    }
    else if ([@"setCustomDimension03" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString *dimension = args[@"dimension"];
        if(dimension == nil)
        {
            dimension = @"";
        }
        [GameAnalytics setCustomDimension03:dimension];
    }
    else if ([@"setGlobalCustomEventFields" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSDictionary *fields_dict = nil;
        if (args[@"customFields"])
        {
            fields_dict = [NSJSONSerialization JSONObjectWithData:[args[@"customFields"] dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:nil];
        }
        [GameAnalytics setGlobalCustomEventFields:fields_dict];
    }
    else if ([@"startSession" isEqualToString:call.method])
    {
        [GameAnalytics startSession];
    }
    else if ([@"endSession" isEqualToString:call.method])
    {
        [GameAnalytics endSession];
    }
    else if ([@"getRemoteConfigsValueAsString" isEqualToString:call.method])
    {
        NSMutableDictionary* args = call.arguments;
        NSString* key = args[@"key"];
        NSString* defaultValue = @"";

        NSString* returnValue = nil;
        if(args[@"defaultValue"])
        {
            defaultValue = args[@"defaultValue"];
            returnValue = [GameAnalytics getRemoteConfigsValueAsString:key defaultValue:defaultValue];
        }
        else
        {
            returnValue = [GameAnalytics getRemoteConfigsValueAsString:key];
        }
        result(returnValue);
    }
    else if ([@"isRemoteConfigsReady" isEqualToString:call.method])
    {
        result(@([GameAnalytics isRemoteConfigsReady]));
    }
    else if ([@"getRemoteConfigsContentAsString" isEqualToString:call.method])
    {
        result([GameAnalytics getRemoteConfigsContentAsString]);
    }
    else if ([@"getRemoteConfigsContentAsString" isEqualToString:call.method])
    {
        result([GameAnalytics getRemoteConfigsContentAsString]);
    }
    else if ([@"getABTestingId" isEqualToString:call.method])
    {
        result([GameAnalytics getABTestingId]);
    }
    else if ([@"getABTestingVariantId" isEqualToString:call.method])
    {
        result([GameAnalytics getABTestingVariantId]);
    }
    else
    {
        result(FlutterMethodNotImplemented);
    }
}

@end
